### PR TITLE
Move of function's local variables to registers

### PR DESCRIPTION
**Pull request by ruben-ayrapetyan**
88cd22f3728acda439e0142fd119399e45e74181:

Native build on RPi2

                              Benchmark |         RSS<br>(+ is better) |        Perf<br>(+ is better) |
                               --------- |                          --- |                         ---- |
                              3d-cube.js |          168->   160 (4.762) |    5.84133->4.13067 (29.285) | 
                  access-binary-trees.js |           96->    96 (0.000) |      3.13733-> 3.104 (1.062) | 
                      access-fannkuch.js |           52->    52 (0.000) |    14.3613->10.6413 (25.903) | 
                         access-nbody.js |          76->    68 (10.526) |    5.91733->4.91333 (16.967) | 
             bitops-3bit-bits-in-byte.js |           36->    36 (0.000) |    4.69333->3.64133 (22.415) | 
                  bitops-bits-in-byte.js |           36->    36 (0.000) |      6.728->4.96667 (26.179) | 
                   bitops-bitwise-and.js |           28->    28 (0.000) |     4.20533-> 4.584 (-9.005) | 
                controlflow-recursive.js |          232->   232 (0.000) |        3.72-> 3.728 (-0.215) | 
                    date-format-xparb.js |          112->   112 (0.000) |      3.42533-> 3.328 (2.841) | 
                          math-cordic.js |           48->    48 (0.000) |      7.436->5.27733 (29.030) | 
                    math-partial-sums.js |           40->    40 (0.000) |        2.908->  2.82 (3.026) | 
                   math-spectral-norm.js |           56->    56 (0.000) |     3.89867->3.54133 (9.166) | 
                         string-fasta.js |           60->    56 (6.667) |      28.3627->28.036 (1.152) | 
                         Geometric mean: |       RSS reduction: 1.7462% |            Speed up: 13.078% |

<br>
Arndale

Benchmark | Rss | Perf
-------- | -------- | --------
                                ./sunspider-1.0.2/3d-cube.js |    168 ->    164 (2.381%) |  3.142 ->  2.062 (34.373%) | 
                    ./sunspider-1.0.2/access-binary-trees.js |     96 ->     96 (0.000%) |  1.838 ->  1.822 (0.871%) | 
                        ./sunspider-1.0.2/access-fannkuch.js |     44 ->     44 (0.000%) |  7.305 ->  5.212 (28.652%) | 
                           ./sunspider-1.0.2/access-nbody.js |     76 ->     64 (15.790%) |  3.173 ->  2.578 (18.752%) | 
               ./sunspider-1.0.2/bitops-3bit-bits-in-byte.js |     36 ->     36 (0.000%) |  2.249 ->  1.589 (29.346%) | 
                    ./sunspider-1.0.2/bitops-bits-in-byte.js |     36 ->     36 (0.000%) |  3.357 ->    2.1 (37.444%) | 
                     ./sunspider-1.0.2/bitops-bitwise-and.js |     28 ->     28 (0.000%) |  1.947 ->  1.981 (-1.746%) | 
                  ./sunspider-1.0.2/controlflow-recursive.js |    224 ->    224 (0.000%) |  2.017 ->  1.994 (1.140%) | 
                      ./sunspider-1.0.2/date-format-xparb.js |    112 ->    112 (0.000%) |  2.002 ->  1.981 (1.049%) | 
                            ./sunspider-1.0.2/math-cordic.js |     44 ->     44 (0.000%) |  3.916 ->  2.435 (37.819%) | 
                      ./sunspider-1.0.2/math-partial-sums.js |     40 ->     40 (0.000%) |  1.552 ->  1.425 (8.183%) | 
                     ./sunspider-1.0.2/math-spectral-norm.js |     56 ->     52 (7.143%) |  1.952 ->  1.753 (10.195%) | 
                           ./sunspider-1.0.2/string-fasta.js |     56 ->     52 (7.143%) | 15.689 -> 15.517 (1.096%) | 
                                              Geometric mean |                          2.6127% |                         17.3387%

<br>
Previous revisions:
Native build on RPi2

                               Benchmark |         RSS<br>(+ is better) |        Perf<br>(+ is better) |
                               --------- |                          --- |                         ---- |
                              3d-cube.js |          168->   160 (4.762) |       5.812-> 4.096 (29.525) | 
                  access-binary-trees.js |           96->    96 (0.000) |         3.15-> 3.126 (0.762) | 
                      access-fannkuch.js |           52->    52 (0.000) |      14.236->10.244 (28.042) | 
                         access-nbody.js |          76->    68 (10.526) |        5.87-> 4.822 (17.854) | 
             bitops-3bit-bits-in-byte.js |           36->    36 (0.000) |       4.688-> 3.434 (26.749) | 
                  bitops-bits-in-byte.js |           36->    36 (0.000) |        6.85->   4.6 (32.847) | 
                   bitops-bitwise-and.js |           28->    28 (0.000) |       4.186-> 4.406 (-5.256) | 
                controlflow-recursive.js |          232->   232 (0.000) |       3.704-> 3.764 (-1.620) | 
                    date-format-xparb.js |          112->   112 (0.000) |        3.392-> 3.332 (1.769) | 
                          math-cordic.js |           48->    48 (0.000) |        7.42-> 5.014 (32.426) | 
                    math-partial-sums.js |           40->    40 (0.000) |        2.922->  2.75 (5.886) | 
                   math-spectral-norm.js |           56->    56 (0.000) |       3.968-> 3.504 (11.694) | 
                         string-fasta.js |           60->    56 (6.667) |       28.156->28.048 (0.384) | 
                         Geometric mean: |       RSS reduction: 1.7462% |           Speed up: 15.0745% |

<br>

Arndale

Benchmark | Rss | Perf
-------- | -------- | --------
./sunspider-1.0.2/3d-cube.js | 168 -> 164 (2.381%) | 3.089 -> 2.028 (34.348%) | 
./sunspider-1.0.2/access-binary-trees.js | 96 -> 96 (0.000%) | 1.865 -> 1.763 (5.469%) | 
./sunspider-1.0.2/access-fannkuch.js | 48 -> 44 (8.333%) | 6.981 -> 4.735 (32.173%) | 
./sunspider-1.0.2/access-nbody.js | 76 -> 64 (15.790%) | 3.222 -> 2.459 (23.681%) | 
./sunspider-1.0.2/bitops-3bit-bits-in-byte.js | 36 -> 36 (0.000%) | 2.306 -> 1.54 (33.218%) | 
./sunspider-1.0.2/bitops-bits-in-byte.js | 36 -> 36 (0.000%) | 3.329 -> 1.979 (40.553%) | 
./sunspider-1.0.2/bitops-bitwise-and.js | 28 -> 28 (0.000%) | 1.851 -> 1.943 (-4.970%) | 
./sunspider-1.0.2/controlflow-recursive.js | 224 -> 224 (0.000%) | 2.032 -> 1.855 (8.711%) | 
./sunspider-1.0.2/crypto-sha1.js | 164 -> 160 (2.439%) | 58.129 -> 58.728 (-1.030%) | 
./sunspider-1.0.2/date-format-xparb.js | 112 -> 112 (0.000%) | 2.005 -> 1.96 (2.244%) | 
./sunspider-1.0.2/math-cordic.js | 48 -> 44 (8.333%) | 3.882 -> 2.344 (39.619%) | 
./sunspider-1.0.2/math-partial-sums.js | 40 -> 40 (0.000%) | 1.541 -> 1.422 (7.722%) | 
./sunspider-1.0.2/math-spectral-norm.js | 56 -> 52 (7.143%) | 1.947 -> 1.68 (13.713%) | 
./sunspider-1.0.2/string-fasta.js | 56 -> 52 (7.143%) | 15.492 -> 15.374 (0.762%) | 
Geometric mean | 3.8035% | 18.4444%

### CODE REVIEW

File: jerry-core/parser/js/opcodes-dumper.cpp [hunk 1]
```diff
@@ -19,7 +19,7 @@
 #include "stack.h"
 #include "jsp-early-error.h"
 
-static idx_t temp_name, max_temp_name;
+static idx_t temp_name, max_temp_name, max_temp_name_for_local_var;
```
Comment (sand1k) [inline jerry-core/parser/js/opcodes-dumper.cpp]: It seems, that these names are not self-descriptive enough. Maybe it is better to name them:
`temp_reg, mar_temp_reg, max_temp_reg_for_local_var`?
And comments are needed.

File: jerry-core/parser/js/scopes-tree.h [hunk 2]
```diff
@@ -49,14 +59,24 @@ typedef struct
...
+                          *   (not depends on subscopes) */
+  bool ref_eval: 1; /**< flag, indicating that "eval" is used inside the scope
+                          *   (not depends on subscopes) */
+  bool contains_with: 1; /**< flag, indicationg whether 'with' statement is contained in the scope
+                          *   (not depends on subscopes) */
+  bool contains_try: 1; /**< flag, indicationg whether 'try' statement is contained in the scope
+                         *   (not depends on subscopes) */
+  bool contains_delete: 1; /**< flag, indicationg whether 'delete' operator is contained in the scope
+                            *   (not depends on subscopes) */
+  bool decl_functions: 1; /**< flag, indicating that functions are declared inside the scope */
```
Comment (sand1k) [inline jerry-core/parser/js/scopes-tree.h]: Maybe rename this to `contains_function_decls` to be consistent in naming?

**Comment by sand1k** [comment]
Don't we need to add some define, which enables/disables this optimization? It could be useful for debugging.

**Comment by sand1k** [comment]
Great patch!

File: jerry-core/parser/js/opcodes-dumper.cpp (hunk 1 above)
Comment (ruben-ayrapetyan) [inline jerry-core/parser/js/opcodes-dumper.cpp]: Yes, sure. I'll update them.

File: jerry-core/parser/js/scopes-tree.h (hunk 2 above)
Comment (ruben-ayrapetyan) [inline jerry-core/parser/js/scopes-tree.h]: Ok.

**Comment by zherczeg** [comment]
LGTM

**Comment by ruben-ayrapetyan** [comment]
@sand1k, I've rebased the patch and also updated it according to your requests. Please, review.

**Comment by sand1k** [comment]
LGTM.

**Pull request merged**